src/iosrv/sched.c

```diff
@@ -0,0 +1,539 @@
+		D_ERROR("XS(%d) poped NULL unit for ABT pool(%d)\n",
+			dx->dx_xs_id, pool_idx);
+
+	cycle->sc_age_net++;
+	cycle->sc_age_nvme++;
+	cycle->sc_ults_cnt[pool_idx] -= 1;
+	cycle->sc_ults_tot -= 1;
+
+	return unit;
+}
+
+static void
+sched_start_cycle(struct sched_data *data, ABT_pool *pools)
+{
+	struct dss_xstream	*dx = data->sd_dx;
+	struct sched_cycle	*cycle = &data->sd_cycle;
+	size_t			 cnt;
+	uint32_t		 limit = 0;
+	bool			 space_pressure = false;
+	int			 i, ret;
+
+	D_ASSERT(cycle->sc_new_cycle == 1);
+	D_ASSERT(cycle->sc_cycle_started == 0);
+	D_ASSERT(cycle->sc_ults_tot == 0);
+
+	cycle->sc_new_cycle = 0;
+	cycle->sc_cycle_started = 1;
+
+	/* Get number of ULTS for each ABT pool */
+	for (i = DSS_POOL_IO; i < DSS_POOL_CNT; i++) {
+		D_ASSERT(cycle->sc_ults_cnt[i] == 0);
+
+		ret = ABT_pool_get_size(pools[i], &cnt);
+		if (ret != ABT_SUCCESS) {
+			D_ERROR("XS(%d) get ABT pool(%d) size error: %d\n",
+				dx->dx_xs_id, i, ret);
+			cnt = 0;
+		}
+
+		cycle->sc_ults_cnt[i] = cnt;
+		cycle->sc_ults_tot += cycle->sc_ults_cnt[i];
+
+		if (sched_throttle[i] > 0 && cycle->sc_ults_cnt[i] > 1)
+			limit = 1;
+	}
+
+	/* No throttling for helper xstream so far */
+	if (!dx->dx_main_xs)
+		return;
+
+	if (!limit && !space_pressure)
+		return;
+
+	/*
+	 * Throttle the ABT pools which have throttle setting.
+	 * TODO: If it's under space pressure, throttle IO pool.
+	 */
+	for (i = DSS_POOL_IO; i < DSS_POOL_CNT; i++) {
+		unsigned int	throttle = sched_throttle[i];
+		int		diff;
+
+		if (sched_throttle[i] == 0)
+			continue;
+
+		/*
+		 * No ULTs from other ABT pools in current cycle, or too few
+		 * ULTs in current cycle.
+		 */
+		if (cycle->sc_ults_cnt[i] == cycle->sc_ults_tot ||
+		    cycle->sc_ults_tot <= cycle->sc_age_net_bound[0])
+			continue;
+
+		D_ASSERT(throttle < 100);
+		limit = max(cycle->sc_ults_tot * throttle / 100, 1);
+		diff = cycle->sc_ults_cnt[i] - limit;
+
+		D_ASSERT(cycle->sc_ults_tot > diff);
+		if (diff > 0 &&
+		    (cycle->sc_ults_tot - diff) > cycle->sc_age_net_bound[0]) {
+			cycle->sc_ults_cnt[i] -= diff;
+			cycle->sc_ults_tot -= diff;
+		}
+	}
+}
+
+static void
+sched_run(ABT_sched sched)
+{
+	struct sched_data	*data;
+	struct sched_cycle	*cycle;
+	struct dss_xstream	*dx;
+	ABT_pool		 pools[DSS_POOL_CNT];
+	ABT_pool		 pool;
+	ABT_unit		 unit;
+	uint32_t		 work_count = 0;
+	int			 i, ret;
+
+	ABT_sched_get_data(sched, (void **)&data);
+	cycle = &data->sd_cycle;
+	dx = data->sd_dx;
+
+	ret = ABT_sched_get_pools(sched, DSS_POOL_CNT, 0, pools);
+	if (ret != ABT_SUCCESS) {
+		D_ERROR("XS(%d) get ABT pools error: %d\n",
+			dx->dx_xs_id, ret);
+		return;
+	}
+
+	while (1) {
+		/* Try to pick network poll ULT */
+		pool = pools[DSS_POOL_NET_POLL];
+		unit = sched_pop_net_poll(data, pool);
+		if (unit != ABT_UNIT_NULL)
+			goto execute;
+
+		/* Try to pick NVMe poll ULT */
+		pool = pools[DSS_POOL_NVME_POLL];
+		unit = sched_pop_nvme_poll(data, pool);
+		if (unit != ABT_UNIT_NULL)
+			goto execute;
```

Comment:
just curious, what is the benefits of doing progress through separate ULT/individual POOL?  can we just call progress explicitly ? or I miss sth?

Reply:
I think call poller explicitly may work as well, but that requires more code changes (at least the dss_srv_handler() have to be revised, I don't want to revise that function to complicate this patch further).

Reply:
ah, calling progress explicitly might cause problem, since it might create ult etc, which is not allowed in sched_run().